Worker transfer data

michalc opened the pull request:
Data passed via the web worker API is now passed as an ArrayBuffer object using the transfer list, so data is copied once from the source location in the heap rather than twice.

The comment regarding the workaround for Chrome bug 169705, where it is slow transferring subarrays has also been removed. This is not a bug, or at least the behaviour is consistent between Firefox and Chrome. See https://code.google.com/p/chromium/issues/detail?id=169705 and benchmarks at http://jsperf.com/web-worker-sub-array-copy.

kripken commented on src/library_browser.js:
```diff
@@ -1268,7 +1268,7 @@ mergeInto(LibraryManager.library, {
       var data = msg.data['data'];
       if (data) {
-        if (!data.byteLength) data = new Uint8Array(data);
+        data = new Uint8Array(data);
```
this is data that just arrived in a message from a worker, so it is already a typed array - why copy it?

michalc replied:
It's not a typed array (or at least it's not meant to be!). In this PR only ArrayBuffers are sent as the `data` property, and `new Uint8Array(<ArrayBuffer>)` doesn't copy the array, it creates a view on the buffer, at least according to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray#Syntax

kripken replied:
ok, now i see.

please add asserts on that `assert(data instanceof ArrayBuffer)`, good for for the check, and makes the code clearer

kripken commented on src/library_browser.js:
```diff
@@ -1305,30 +1305,33 @@ mergeInto(LibraryManager.library, {
       });
       info.awaited++;
     }
+    data = data ? Module.buffer.slice(data, data + size) : 0;
```
need to write `Module['buffer']` so that this is ok under closure compiler advanced.

kripken commented:
Please verify the `browser.test_*worker*` tests pass.

Did you benchmark this across multiple browsers?

michalc commented:
Added the assertions, changed to `Module['buffer']`, and the `browser.test_worker*` tests do pass.

However, when I run the benchmarks (by `python tests/runner.py benchmark`) then I see a lot of `ERROR` messages, but this happens on the incoming branch as well, so I don't think it's my change.

kripken commented:
Looks good, thanks.

It's ok to ignore the benchmark suite. However, did you manually measure the speed before and after this change? We should be sure that it actually speeds things up in multiple browsers.

michalc commented:
I've now done some benchmarking of transferring 8mb arrays from the worker to the main thread, and the results are at http://jsperf.charemza.name/emscripten-transfer-data-from-worker . It doesn't have appeared to make any difference in Chrome or Firefox, and only a sliver of a difference in Safari...

kripken commented:
That's surprising, isn't it? Perhaps something is not going right, in that the browsers aren't able to transfer these objects? Might be worth talking with browser devs on this.

kripken commented:
Closing as this is quite old. Feel free to reopen if still relevant.
